cpp/src/delivery.cpp

```diff
@@ -40,6 +42,7 @@ namespace proton {
 
 delivery::delivery(pn_delivery_t* d): transfer(make_wrapper(d)) {}
 receiver delivery::receiver() const { return make_wrapper<class receiver>(pn_delivery_link(pn_object())); }
+tag delivery::tag() const { return make_wrapper<class tag>(pn_delivery_tag(pn_object())); }
```

Comment:
This does not compile for me too. The reason seems to be that `make_wrapper` expects to get a pointer type, but `pn_delivery_tag` returns the struct directly as a value, not a pointer to it. If I exactly knew right now what you should do here, I'd tell you...

Reply:
@jiridanek I added an additional template in `cpp/src/proton_bits.hpp` and the error went away. Now I am able to build but I am not sure if we can do it like this, WDYT?
Meanwhile, I am trying writing tests so that we can know if it is working fine or not.

Reply:
I started seeing error after make clean :( will take a look again.

```In file included from /home/rakhi/qpid-proton/cpp/src/delivery.cpp:28:
/home/rakhi/qpid-proton/cpp/src/proton_bits.hpp: In instantiation of ‘U proton::make_wrapper(typename proton::internal::wrapped<T>::type) [with U = proton::tag; typename proton::internal::wrapped<T>::type = pn_bytes_t]’:
/home/rakhi/qpid-proton/cpp/src/delivery.cpp:45:88:   required from here
/home/rakhi/qpid-proton/cpp/src/proton_bits.hpp:159:91: error: cannot convert ‘proton::internal::wrapped<proton::tag>::type’ {aka ‘pn_bytes_t’} to ‘proton::internal::wrapped<proton::tag>::type*’ {aka ‘pn_bytes_t*’}
  159 | U make_wrapper(typename internal::wrapped<U>::type t) { return internal::factory<U>::wrap(t); }
      |                                                                                           ^
      |                                                                                           |
      |                                                                                           proton::internal::wrapped<proton::tag>::type {aka pn_bytes_t}
```

Reply:
Yeah, that's the one I saw too. 

```
cannot convert ‘proton::internal::wrapped<proton::tag>::type’ {aka ‘pn_bytes_t’} to ‘proton::internal::wrapped<proton::tag>::type*’ {aka ‘pn_bytes_t*’}
```

Reply:
@ssorj I think I now see what you meant, the `pn_delivery_tag` is a type alias for `pn_bytes_t`, and the reasonable types for in in C++ are std::string or std::vector<uint8>. So, one of the two following functions should be used instead of `make_wrapper`

https://github.com/apache/qpid-proton/blob/8ddf5399013b02f1fd13bda3fee7886bd48a98be/cpp/src/types_internal.hpp#L58-L59

Both make a copy of the data, which simplifies things.

Reply:
Ohh so I was supposed to do it like this, Thanks Jiri for pointing this out!! Updated the PR. Now working on writing the test for delivery tag accessor and cleaning the PR :)

Reply:
I added a test `cpp/src/delivery_test.cpp` for delivery tag accessor but it looks too simple. Should I add a mutator for full testing? WDYT? @jiridanek

Reply:
I think so. If you look at the tests that @pjfawcett wrote for these two enhancements https://github.com/apache/qpid-proton/pulls?q=is%3Apr+author%3Apjfawcett , you'll see that there are tests that set something when a message is being sent/link is being created/... and then the tests check it on the receiving end. **I think that this would be a good test in addition to what you have already. It demonstrates that an end-user of the library is actually able to use the tags.**

As you know, in AMQP you can set a tag when you are sending a message and you will see the tag in the receiver when your program decides whether to accept the message, and then again the tag is given back to the sender in the delivery after the message has been disposed of.

Looking at the C++ binding, I see that when I send a message, I won't be able to set my own tag, because the binding sets it for me (and tries moderately hard to give unique tag). For example, in the Python binding, there is `def send(self, sender, tag=None):` method in `message.py` so there I can set my own tag.

https://github.com/apache/qpid-proton/blob/8ddf5399013b02f1fd13bda3fee7886bd48a98be/cpp/src/sender.cpp#L62-L72

When I receive the message, I will be able to read the delivery tag in the `on_message` handler method

https://github.com/apache/qpid-proton/blob/8ddf5399013b02f1fd13bda3fee7886bd48a98be/cpp/examples/flow_control.cpp#L181

Finally, when the message is settled in the `on_message` handler of the receiver (using one of the `.accept()`, `reject()`, ... methods, I want to be able to see in the sender the final disposition and the tag. That should cause the handler method `on_tracker_accept`/`on_tracker_reject`/... to be called. **Problem is, I cannot access the delivery and therefore the delivery from that method, at least I did not see a way!**

https://github.com/apache/qpid-proton/blob/8ddf5399013b02f1fd13bda3fee7886bd48a98be/cpp/examples/direct_send.cpp#L76

Reply:
@ssorj For your example, is is necessary to set custom delivery tags when sending messages, or is it ok to have the tags generated by the binding, and then make them available read-only? Do you need to be absolutely sure that delivery tags are unique?

Reply:
I would think it would be fine for a binding to generate one by default, with then perhaps a separate ability to set it yourself if you wanted (many probably wont need to). 

The set of tags needs to be unique for deliveries considered unsettled by either end of a given link. You can reuse tags from previously settled deliveries if you like. Different links can use the same tags at the same time.

Reply:
@jiridanek @gemmellr  Thanks for the explanation.
Fow now added the default delivery tag argument in `sender::send()`
`PN_CPP_EXTERN tracker send(const message &m,const binary &tag = binary());`
WDYT?

Meanwhile working on writing tests and code formatting.

Reply:
The `send()` method makes sense to me. If the API it good enough for Python, it should make some sense for C++ too ;P

I'll add few notes to the code. It is not something to be addressed right away. Those are just improvements I thought of. I agree that writing a test should be the thing to do next.

Reply:
Looks like the PR is basically there. Just fix the ThreadSanitizer warning and decide if you want to tackle the `tracker` here or in a new PR after this one (I'd say now add the handler methods and assert what you can; leave updating the tracker API (to make tag accessible there) for a new task).

Then we can ask one of the Proton C++ devs to review it.

Reply:
Fixed the `ThreadSanitizer warning` and added the `on_delivery_settle()` method. I was thinking of adding the tracker methods in a new PR.